### PR TITLE
Hitshape

### PR DESCRIPTION
**Pull request by huwpascoe**
Final part of https://github.com/OpenRA/OpenRA/issues/6820, new shapes added: Circle, Capsule, Rectangle.

demonstration with arbitrary values:

```
MCV:
    Health:
        HP: 600
        Shape: Capsule
            Radius: 426
            Vertices: 0, -512, 0, 512
```

hit-a
hit-b

```
4TNK:
    Health:
        HP: 900
        Shape: Rectangle
            Vertices: -512, -512, 512, 512
```

```
AFLD:
    Health:
        HP: 1000
        Shape: Rectangle
            Vertices: -1280,-384, 1280, 1024
```

hit-c

### CODE REVIEW

**Comment by huwpascoe** [comment]
This PR not only detects a direct hit, but also computes damage falloff

Like this (visualization from stack overflow)

File: OpenRA.Mods.Common/Traits/Health.cs [hunk 1]
```diff
@@ -17,14 +18,27 @@ public class HealthInfo : ITraitInfo, UsesInit<HealthInit>
...
+		public readonly int2[] Vertices = null;
 
-		[Desc("Trigger interfaces such as AnnounceOnKill?")]
-		public readonly bool NotifyAppliedDamage = true;
+		public IHitShape ShapeInstance { get; private set; }
 
-		public virtual object Create(ActorInitializer init) { return new Health(init, this); }
+		public virtual object Create(ActorInitializer init)
+		{
+			switch (Shape)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Health.cs]: Use `IRulesetLoaded` / `RulesetLoaded` instead of putting this here (where it is invoked every time a new actor is created).

File: OpenRA.Mods.Common/HitShapes/IHitShape.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+using OpenRA.Graphics;
+
+namespace OpenRA.Mods.Common.HitShapes
+{
+	public enum HitShape { Circle, Rectangle, Capsule }
+
+	public interface IHitShape
+	{
+		int OuterRadius { get; }
+		int DistanceFromEdge(int2 p);
```
Comment (pchote) [inline OpenRA.Mods.Common/HitShapes/IHitShape.cs]: This should take a `WVec` and return a `WDist` to cleanly integrate with the rest of our world coordinates.

File: OpenRA.Mods.Common/HitShapes/IHitShape.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+using OpenRA.Graphics;
+
+namespace OpenRA.Mods.Common.HitShapes
+{
+	public enum HitShape { Circle, Rectangle, Capsule }
+
+	public interface IHitShape
+	{
+		int OuterRadius { get; }
```
Comment (pchote) [inline OpenRA.Mods.Common/HitShapes/IHitShape.cs]: `WDist` here too.

File: OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs [hunk 4]
```diff
@@ -70,7 +70,8 @@ public static bool GetDirectHit(World world, CPos cell, WPos pos)
 					continue;
 
 				// If the impact position is within any actor's health radius, we have a direct hit
-				if ((unit.CenterPosition - pos).LengthSquared <= healthInfo.Radius.LengthSquared)
+				var or = healthInfo.ShapeInstance.OuterRadius;
```
Comment (pchote) [inline OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs]: You can then use `.LengthSquared` instead of `or*or`.

File: OpenRA.Mods.Common/HitShapes/Capsule.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+			this.radius = radius;
+
+			outerRadius = radius + Math.Max(v.Length, w.Length);
+		}
+
+		public int DistanceFromEdge(int2 p)
+		{
+			var vec = w - v;
+
+			var lenSq = vec.LengthSquared >> 10;
```
Comment (pchote) [inline OpenRA.Mods.Common/HitShapes/Capsule.cs]: Just write `/ 1024` and let the compiler do its job.

File: OpenRA.Mods.Common/Traits/Health.cs [hunk 6]
```diff
@@ -17,14 +18,27 @@ public class HealthInfo : ITraitInfo, UsesInit<HealthInit>
...
 		[Desc("HitPoints")]
 		public readonly int HP = 0;
+		[Desc("Trigger interfaces such as AnnounceOnKill?")]
+		public readonly bool NotifyAppliedDamage = true;
 
-		[Desc("Physical size of the unit used for damage calculations. Impacts within this radius apply full damage.")]
+		[Desc("Shape of the unit used to calculate warhead damage")]
+		public readonly HitShape Shape = HitShape.Circle;
 		public readonly WDist Radius = new WDist(426);
+		public readonly int2[] Vertices = null;
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Health.cs]: Can we get descriptions for these new fields?

File: OpenRA.Mods.Common/Traits/Health.cs (hunk 1 above)
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Health.cs]: Actually, I would have annotate `ShapeInstance` with a custom `[LoadUsing]` and then make the vertices / other info a sub-tree of the field.  You can then drop the Vertices definition here, where it doesn't really belong.

**Comment by pchote** [comment]
I haven't looked closely, but it looks mostly sensible overall.  Cool!

File: OpenRA.Mods.Common/Traits/Health.cs (hunk 1 above)
Comment (huwpascoe) [inline OpenRA.Mods.Common/Traits/Health.cs]: Yay I wondered if there was something to do that

File: OpenRA.Mods.Common/Traits/Health.cs [hunk 7]
```diff
@@ -8,21 +8,36 @@
...
 
-		[Desc("Trigger interfaces such as AnnounceOnKill?")]
-		public readonly bool NotifyAppliedDamage = true;
+		public IHitShape ShapeInstance { get; private set; }
+
+		public void RulesetLoaded(Ruleset rules, ActorInfo info)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Health.cs]: (copied from the previous tree, where it is now marked as outdated):



> Actually, I would annotate ShapeInstance with a custom [LoadUsing] and then make the vertices / other info a sub-tree of the field. You can then drop the Vertices from the parent Health, where it doesn't really belong.

File: OpenRA.Mods.Common/Traits/Health.cs [hunk 8]
```diff
@@ -17,13 +19,36 @@ public class HealthInfo : ITraitInfo, UsesInit<HealthInit>
...
+			if (nodes.ContainsKey("Radius"))
+				radius = FieldLoader.GetValue<int>("Radius", nodes["Radius"].Value);
+
+			int2[] verts = null;
+			if (nodes.ContainsKey("Vertices"))
+				verts = FieldLoader.GetValue<int2[]>("Vertices", nodes["Vertices"].Value);
+
+			switch (shape)
+			{
+				case "Rectangle": return new Rectangle(verts[0], verts[1]);
```
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Health.cs]: This should include:
```c#
if (Vertices == null || Vertices.Length != 2)
	throw new YamlException("Rectangles require top-left and bottom-right vertices");
```

File: OpenRA.Mods.Common/Traits/Health.cs [hunk 9]
```diff
@@ -17,13 +19,36 @@ public class HealthInfo : ITraitInfo, UsesInit<HealthInit>
...
+				case "Capsule": return new Capsule(verts[0], verts[1], radius);
```
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Health.cs]: This should include:
```c#
if (Vertices == null || Vertices.Length != 2)
	throw new YamlException("Capsules require top-left and bottom-right inner rectangle vertices");
```

**Comment by atlimit8** [comment]
You replaced RulesetLoaded?
Why?

**Comment by atlimit8** [comment]
I wrote it for situations like this.

**Comment by huwpascoe** [comment]
@atlimit8 to quote pchote

"Actually, I would annotate ShapeInstance with a custom [LoadUsing] and then make the vertices / other info a sub-tree of the field. You can then drop the Vertices from the parent Health, where it doesn't really belong."

I applied that change. Which should I go for then, IRulesetLoaded or LoadUsing?

File: OpenRA.Mods.Common/Traits/Health.cs [hunk 10]
```diff
@@ -17,13 +19,36 @@ public class HealthInfo : ITraitInfo, UsesInit<HealthInit>
...
+		static object LoadShape(MiniYaml yaml)
+		{
+			var nodes = yaml.ToDictionary();
+
+			string shape = "Circle";
+			if (nodes.ContainsKey("Shape"))
+				shape = FieldLoader.GetValue<string>("Shape", nodes["Shape"].Value);
+
+			int radius = 426;
+			if (nodes.ContainsKey("Radius"))
```
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Health.cs]: Is this what it is supposed to parse?
```Yaml
MCV:
    Health:
        HP: 600
        Shape: Capsule
            Radius: 426
            Vertices: 0, -512, 0, 512
```
If so, then `nodes = nodes["Shape"].Value`.

**Comment by atlimit8** [comment]
LoadUsing should be used for
```Yaml
MCV:
    Health:
        HP: 600
        Shape: Capsule
            Radius: 426
            Vertices: 0, -512, 0, 512
```
because recursive MinYaml loading is not supported (especially custom object creation dispatch).

**Comment by huwpascoe** [comment]
@atlimit8 Is this better?

File: OpenRA.Mods.Common/Traits/Health.cs [hunk 11]
```diff
@@ -17,13 +19,40 @@ public class HealthInfo : ITraitInfo, UsesInit<HealthInit>
...
+				if (nodes.ContainsKey("Radius"))
+					radius = FieldLoader.GetValue<int>("Radius", nodes["Radius"].Value);
+
+				if (nodes.ContainsKey("Vertices"))
+					verts = FieldLoader.GetValue<int2[]>("Vertices", nodes["Vertices"].Value);
+			}
+
+			switch (shape)
+			{
+				case "Rectangle": return new Rectangle(verts[0], verts[1]);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Health.cs]: If you use reflection to create these and have the constructor accept a list of miniyaml nodes then you can avoid hardcoding the types, and modders could add new ones.

File: OpenRA.Mods.Common/Traits/Health.cs [hunk 12]
```diff
@@ -17,13 +19,40 @@ public class HealthInfo : ITraitInfo, UsesInit<HealthInit>
...
+
+			var shapeNode = yaml.Nodes.Find(n => n.Key == "Shape");
+			if (shapeNode != null)
+			{
+				shape = shapeNode.Value.Value;
+
+				var nodes = shapeNode.Value.ToDictionary();
+
+				if (nodes.ContainsKey("Radius"))
+					radius = FieldLoader.GetValue<int>("Radius", nodes["Radius"].Value);
```
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Health.cs]: `WDist` type should be kept here instead of loading an `int` for clarity of the length of the radius.

File: OpenRA.Mods.Common/Traits/Health.cs (hunk 11 above)
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Health.cs]: This should include:
```c#
if (Vertices == null || Vertices.Length != 2)
	throw new YamlException("Rectangel requires top-left and bottom-right vertices");
```
- or better -
Use @pchote's suggestion and put these checks in the constructors.

**Comment by atlimit8** [comment]
Put those two checks in the cases or in the constructors.

File: OpenRA.Mods.Common/Traits/Health.cs (hunk 11 above)
Comment (huwpascoe) [inline OpenRA.Mods.Common/Traits/Health.cs]: Can you show me an example? Reflection isn't my strong point

Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Health.cs]: Add `void Loaded()` to `IHitShape`:
```c#
	public interface IHitShape
	{
		WDist OuterRadius { get; }
		WDist DistanceFromEdge(WVec p);
		void DrawCombatOverlay(WorldRenderer wr, LineRenderer wlr, WPos center, WRot angle);
		void Loaded(); // Called once loaded from MiniYaml
	}
```

---
Modify hit shapes to something like:
```c#
	public class CircleHitShape : IHitShape
	{
		public WDist OuterRadius { get { return Radius; } }
		public readonly WDist Radius = new WDist(426);

		public void Loaded() { }

		...
	}
```
```c#
	public class CapsuleHitShape : IHitShape
	{
		public WDist OuterRadius { get { return outerRadius; } }
		[FieldLoader.required]
		public readonly int2 TopLeft;
		[FieldLoader.required]
		public readonly int2 BottomRight;
		public readonly WDist Radius = new WDist(426);
		WDist outerRadius;

		public void Loaded()
		{
			outerRadius = new WDist(radius + Math.Max(v.Length, w.Length));
		}

		...
	}
```
```c#
	public class RectangleHitShape : IHitShape
	{
		public WDist OuterRadius { get { return outerRadius; } }
		[FieldLoader.required]
		public readonly int2 Size;
		public readonly int2 Center = new int2(0, 0);
		public readonly WDist Radius = new WDist(426);
		WDist outerRadius;
		readonly WVec[] combatOverlayVerts;

		public void Loaded()
		{
			if (Size.X <= 0 || Size.Y <= 0)
				throw new YamlException("Size must be positive values");

			var tl = Center - Size / 2;
			var br = Center + Size / 2;
			combatOverlayVerts = new WVec[]
			{
				new WVec(tl.X, tl.Y, 0),
				new WVec(br.X, tl.Y, 0),
				new WVec(br.X, br.Y, 0),
				new WVec(tl.X, br.Y, 0),
				new WVec(tl.X, tl.Y, 0)
			};

			outerRadius = new WDist(Math.Max(tl.Length, br.Length));
		}

		...
	}
```
Note: each shape can have its own MiniYaml fields

---
Loader code:
```c#
		static object LoadShape(MiniYaml yaml)
		{
			var shapeNode = yaml.Nodes.Find(n => n.Key == "Shape");
			var shape = shapeNode == null || string.IsNullOrEmpty(shapeNode.Value) ? "Circle" : shapeNode.Value;
			var ret = Game.CreateObject<IHitShape>(shape + "HitShape");
			if (shapeNode != null)
			{
				try
				{
					FieldLoader.Load(ret, shapeNode);
				}
				catch (YamlException ex)
				{
					throw new YamlException("Shape " + shape + ": " + ex.Message));
				}
			}

			ret.Loaded();
			return ret;
		}
```

File: OpenRA.Mods.Common/HitShapes/IHitShape.cs [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using OpenRA.Graphics;
+
+namespace OpenRA.Mods.Common.HitShapes
+{
+	public interface IHitShape
```
Comment (pchote) [inline OpenRA.Mods.Common/HitShapes/IHitShape.cs]: Can you please also add implementations for line-vs-shape intersections so that this can play nicely with #9357?

File: OpenRA.Mods.Common/HitShapes/Capsule.cs [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+			this.w = w;
+			this.radius = radius;
+
+			outerRadius = new WDist(radius + Math.Max(v.Length, w.Length));
+		}
+
+		public WDist DistanceFromEdge(WVec pv)
+		{
+			var p = new int2(pv.X, pv.Y);
+			var vec = w - v;
```
Comment (pchote) [inline OpenRA.Mods.Common/HitShapes/Capsule.cs]: `w` and `v` are both constant, so you can calculate `vec` once in the constructor.

**Comment by huwpascoe** [comment]
https://github.com/OpenRA/OpenRA/pull/9800#discussion_r43492030
I've just had a short attempt at doing it but ray->rectangle, ray->capsule is a lot more difficult so this will take some time.

@atlimit8 Thanks, it works well. I'll have the next PR soon

File: OpenRA.Mods.Common/HitShapes/IHitShape.cs (hunk 13 above)
Comment (UberWaffe) [inline OpenRA.Mods.Common/HitShapes/IHitShape.cs]: tl;dr: ```WDist DistanceFromEdge(WVec p);``` might already suit the needs of #9357.

!tl;dr
With regards to #9357, does the implementations of the line below ```WDist DistanceFromEdge(WVec p);``` give you the distance from the actor center position to the edge of its shape along a vector?

If so, in #9357 the [actor width line] (https://github.com/OpenRA/OpenRA/pull/9357/files#diff-5f66c498eea5ea840b8d556a7285788eR49) can be adjusted to use this instead of assuming circle radius (i.e. equal radius in all directions).

Basically, I could use it to get the actor's 'width' along the vector from its center point to the closest point on the line.

**Comment by UberWaffe** [comment]
From the [comment] (https://github.com/OpenRA/OpenRA/pull/9800#discussion_r43573052) above:

@huwpascoe:
tl;dr: WDist DistanceFromEdge(WVec p); might already suit the needs of #9357.

!tl;dr
With regards to #9357, does the implementations of the line below WDist DistanceFromEdge(WVec p); give you the distance from the actor's center position to the edge of its shape along a vector? (it appears to do that)

If so, in #9357 the actor width line can be adjusted to use WDist DistanceFromEdge(WVec p); instead of assuming circle radius (i.e. equal radius in all directions).

Basically, I could use it to get the actor's 'width' along the vector from its center point to the closest point on the line.
The code already has both WPos, so it should be straightforward.

I think the following should be what it ends up as in #9357

**Comment by reaperrr** [comment]
> OpenRA.Mods.Common/HitShapes/Circle.cs(24,7): error CS0414: Warning as Error: The private field `OpenRA.Mods.Common.HitShapes.HitShapeCircle.radiusSq' is assigned but its value is never used

**Comment by huwpascoe** [comment]
@reaperrr Don't worry I just noticed it : P (you'd think visual studio would have picked up on that first)

@UberWaffe If you think you can make do with the dist methods then I'll post as is. I got as far as line/circle and line/rectangle (line/capsule still stuck on) so if it turns out you want those methods later I can post them separately.

File: OpenRA.Mods.Common/Traits/Health.cs [hunk 15]
```diff
@@ -17,13 +19,39 @@ public class HealthInfo : ITraitInfo, UsesInit<HealthInit>
...
+		static object LoadShape(MiniYaml yaml)
+		{
+			IHitShape ret;
+
+			var shapeNode = yaml.Nodes.Find(n => n.Key == "Shape");
+			var shape = shapeNode != null ? shapeNode.Value.Value : string.Empty;
+
+			if (!string.IsNullOrEmpty(shape))
+			{
+				ret = Game.CreateObject<IHitShape>("HitShape" + shape);
```
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Health.cs]: The idiom is to add a suffix instead of a prefix:
```
OpenRA.Game/GameRules/ActorInfo.cs:112:    var info = Game.CreateObject<ITraitInfo>(traitName + "Info");
OpenRA.Game/GameRules/WeaponInfo.cs:78:    var ret = Game.CreateObject<IProjectileInfo>(proj.Value + "Info");
OpenRA.Game/GameRules/WeaponInfo.cs:88:    var ret = Game.CreateObject<IWarhead>(node.Value.Value + "Warhead");
OpenRA.Game/Map/ActorReference.cs:46:      var info = Game.CreateObject<IActorInit>(traitName + "Init");
OpenRA.Game/Widgets/WidgetLoader.cs:77:    return Game.ModData.ObjectCreator.CreateObject<Widget>(widgetType + "Widget", args);

```
It also sounds better.

**Comment by huwpascoe** [comment]
suffix'd "Shape" to match the rest of the project

**Comment by huwpascoe** [comment]
Added unit tests and fixed a bunch of stupid mistakes.

**Comment by huwpascoe** [comment]
Fixed merge conflict

**Comment by huwpascoe** [comment]
Merged with upstream, still passes.

**Comment by penev92** [comment]
You need to rebase onto bleed to get rid of the "Merge" commits.

**Comment by huwpascoe** [comment]
lol mybad. Rebased.

**Comment by pchote** [comment]
We're really low on review bandwidth right now, and the time that we do have is mainly focused towards the next release.  It may be a while longer before we can continue reviewing this, but please don't take that as a slight against you personally!

**Comment by huwpascoe** [comment]
No worries!

File: OpenRA.Mods.Common/HitShapes/Circle.cs [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+using OpenRA.Graphics;
+
+namespace OpenRA.Mods.Common.HitShapes
+{
+	public class CircleShape : IHitShape
+	{
+		public WDist OuterRadius { get { return Radius; } }
+
+		[FieldLoader.Require]
+		public readonly WDist Radius = new WDist(426);
```
Comment (penev92) [inline OpenRA.Mods.Common/HitShapes/Circle.cs]: Is it me, or is this 0.416 cells? Please default it to 512 or something.

File: OpenRA.Mods.Common/HitShapes/Circle.cs [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+
+		public CircleShape() { }
+
+		public CircleShape(WDist radius) { Radius = radius; }
+
+		public void Initialize() { }
+
+		public WDist DistanceFromEdge(WVec p)
+		{
+			return new WDist(Math.Max(0, p.Length - Radius.Length));
```
Comment (penev92) [inline OpenRA.Mods.Common/HitShapes/Circle.cs]: I don't get it, why the 0?

File: OpenRA.Mods.Common/HitShapes/Circle.cs (hunk 16 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/HitShapes/Circle.cs]: @penev92: The internal default for health radius was also 426 ( = 10 pixel in RA/TD).

Comment (penev92) [inline OpenRA.Mods.Common/HitShapes/Circle.cs]: Wow. YAY for random defaults...
Never mind then...

File: OpenRA.Mods.Common/HitShapes/Circle.cs (hunk 17 above)
Comment (huwpascoe) [inline OpenRA.Mods.Common/HitShapes/Circle.cs]: It returns the distance relative to the edge of the hitshape, <= 0 means inside the hitshape, and so a direct hit. Since negative distance makes no sense, needs to be clamped to 0.

Comment (penev92) [inline OpenRA.Mods.Common/HitShapes/Circle.cs]: Ah, so you only care for values outside the shape. OK then.

**Comment by huwpascoe** [comment]
fixed build error with bleed

File: OpenRA.Mods.Common/HitShapes/IHitShape.cs [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+	{
+		WDist OuterRadius { get; }
+
+		WDist DistanceFromEdge(WVec p);
```
Comment (pchote) [inline OpenRA.Mods.Common/HitShapes/IHitShape.cs]: This requires `p` to be already rotated into the local coordinate system of the victim (see SpreadDamageWarhead:L69) which breaks the abstraction a bit, IMO.

How about we make this `WDist DistanceFromEdge(Actor self, WPos position)`, and then calculate the vector, rotate it, and return the distance all internally.  The circle shape can then just return the vector distance directly.

Comment (huwpascoe) [inline OpenRA.Mods.Common/HitShapes/IHitShape.cs]: Seems kinda repetitive, but I'll give it a try

**Comment by huwpascoe** [comment]
Okay replacing the vectors with Actor is a lot cleaner after all

File: OpenRA.Test/OpenRA.Test.csproj [hunk 19]
```diff
@@ -58,6 +59,11 @@
       <Name>OpenRA.Game</Name>
       <Private>False</Private>
     </ProjectReference>
+    <ProjectReference Include="..\OpenRA.Mods.Common\OpenRA.Mods.Common.csproj">
```
Comment (UberWaffe) [inline OpenRA.Test/OpenRA.Test.csproj]: Are these additions in `csproj` needed?

**Comment by UberWaffe** [comment]
Did a quick test in TS.
The rectangle hitshape does not appear to take the isometric perspective into account.

File: OpenRA.Test/OpenRA.Test.csproj (hunk 19 above)
Comment (pchote) [inline OpenRA.Test/OpenRA.Test.csproj]: You'll also need to fix the makefile:
```
paul@Mjolnir:OpenRA+pr/9800$ make nunit
CSC OpenRA.Test.dll
OpenRA.Test//OpenRA.Game/ShapeTest.cs(13,14): error CS0234: The type or namespace name `Mods' does not exist in the namespace `OpenRA'. Are you missing an assembly reference?
OpenRA.Test//OpenRA.Game/ShapeTest.cs(20,3): error CS0246: The type or namespace name `IHitShape' could not be found. Are you missing an assembly reference?
Compilation failed: 2 error(s), 0 warnings
```

File: OpenRA.Test/OpenRA.Game/ShapeTest.cs [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System.IO;
+using NUnit.Framework;
+using OpenRA.Mods.Common.HitShapes;
+
+namespace OpenRA.Test
```
Comment (pchote) [inline OpenRA.Test/OpenRA.Game/ShapeTest.cs]: This appears to be in the wrong place.  The file path implies that it is a Game test, but it then depends on things from mods.common.  Can you put this in a new Mods.Common dir instead?

File: OpenRA.Test/OpenRA.Game/ShapeTest.cs [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+			shape.Initialize();
+
+			Assert.That(shape.DistanceFromEdge(new WVec(100, 100, 0)).Length,
+				Is.EqualTo(0));
+
+			Assert.That(shape.DistanceFromEdge(new WVec(1000, 0, 0)).Length,
+				Is.EqualTo(0));
+
+			Assert.That(shape.DistanceFromEdge(new WVec(2000, 2000, 0)).Length,
+				Is.InRange(1593, 1595));
```
Comment (pchote) [inline OpenRA.Test/OpenRA.Game/ShapeTest.cs]: Why not `Is.Equals(1594)` and so on everywhere else?

**Comment by pchote** [comment]
Code looks good to me aside from those remaining test nits.  I haven't had a chance to test this ingame.

File: OpenRA.Test/OpenRA.Game/ShapeTest.cs (hunk 21 above)
Comment (huwpascoe) [inline OpenRA.Test/OpenRA.Game/ShapeTest.cs]: The result is fractional so there's going to be a margin of error with these integer based maths.

Comment (pchote) [inline OpenRA.Test/OpenRA.Game/ShapeTest.cs]: It is deterministic with well defined rules: you will always get the same result.

File: OpenRA.Test/OpenRA.Test.csproj (hunk 19 above)
Comment (huwpascoe) [inline OpenRA.Test/OpenRA.Test.csproj]: I can't reproduce in windows

Comment (pchote) [inline OpenRA.Test/OpenRA.Test.csproj]: Windows doesn't use the makefile.

**Comment by huwpascoe** [comment]
I've made changes (although I can't test the makefile works)

**Comment by reaperrr** [comment]
Needs a rebase.

**Comment by pchote** [comment]
Unit tests now work.

**Comment by huwpascoe** [comment]
Rebased

File: OpenRA.Test/OpenRA.Mods.Common/ShapeTest.cs [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+			Assert.That(shape.DistanceFromEdge(new WVec(2000, 2000, 0)).Length,
+				Is.InRange(1593, 1595));
+
+			Assert.That(new CircleShape(new WDist(73))
+				.DistanceFromEdge(new WVec(150, -100, 0)).Length,
+				Is.InRange(106, 108));
+
+			Assert.That(new CircleShape(new WDist(55))
+				.DistanceFromEdge(new WVec(30, -45, 0)).Length,
+				Is.InRange(-1, 1));
```
Comment (RoosterDragon) [inline OpenRA.Test/OpenRA.Mods.Common/ShapeTest.cs]: I swear there was an earlier comment about this, but I agree that these tests should specify exact values and not ranges. The output must be deterministic or else we'll get desyncs.

**Comment by RoosterDragon** [comment]
Do we want to tackle actually applying this anywhere in this PR, or is somebody going to follow up with some changes?

Also, needs a rebase.

**Comment by reaperrr** [comment]
> Do we want to tackle actually applying this anywhere in this PR, or is somebody going to follow up with some changes?

I wouldn't mind doing a follow-up if nobody else volunteers. Using rectangular hitboxes for buildings, for example, will likely not only result in a large yaml diff, but also have an effect on balance, so I believe it would be better to tackle that in a separate PR (or multiple ones).

**Comment by huwpascoe** [comment]
Refactored for the new RgbaColorRenderer.

This PR is for the bounty of "Add support for rectangular hitboxes" which I've delivered. Rule changes are another issue entirely. : V

**Comment by reaperrr** [comment]
This PR would currently break any custom health radius (it actually does for infantry on all mods) and needs an upgrade rule that checks whether a custom `Health.Radius` is defined in yaml rules, and if yes, replaces `Radius: xx` with `Shape: Circle` -> `Radius: xx`.

Luckily for you, I had nothing better to do ;)
https://github.com/reaperrr/OpenRA/tree/hitshape-ur

Just pick the last commit.

Other than that and the minor rectangle overlay issue, everything seems to work fine in-game.
:+0.5: for ingame from me once those are adressed, since this is a bit out of my league in terms of reviewing code.

File: OpenRA.Mods.Common/HitShapes/Rectangle.cs [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+			TopLeft = tl;
+			BottomRight = br;
+		}
+
+		public void Initialize()
+		{
+			if (TopLeft.X >= BottomRight.X || TopLeft.Y >= BottomRight.Y)
+				throw new YamlException("TopLeft and BottomRight points are invalid.");
+
+			size = (BottomRight - TopLeft) / 2;
```
Comment (atlimit8) [inline OpenRA.Mods.Common/HitShapes/Rectangle.cs]: nit: This is a quadrant size rather than the actual dimensions. Perhaps `quadrantSize`.

**Comment by huwpascoe** [comment]
@reaperrr Thanks, that saved me a lot of work

Nits picked and bugs fixed.

**Comment by RoosterDragon** [comment]
:+1: Code looks fine. Debug visuals seem fine. Works fine in TS's isometric perspective. Damage is being dealt correctly in accordance with the debug visuals. Tests look fine and also pass.

I have no problem leaving actual rule changes to a later PR.

Revised YAML from the opening post for anybody else doing a test run:
```
MCV:
	Health:
		HP: 600
		Shape: Capsule
            Radius: 426
            PointA: 0, -512
			PointB: 0, 512
```

```
4TNK:
	Health:
		HP: 900
		Shape: Rectangle
            TopLeft: -512, -512
			BottomRight: 512, 512
```

```
AFLD:
	Health:
		HP: 1000
		Shape: Rectangle
            TopLeft: -1280, -384
			BottomRight: 1280, 1024
```

**Comment by abcdefg30** [comment]
Looks good to me, too. :+1: / :white_check_mark:

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/349ef44f49580951edbc855c4827104149b11c6d...61ee467173d19078eb221d21c9a15ba167f9ad89)

**Comment by pchote** [comment]
Unfortunately this broke our unit tests #10325.
